### ISSUE TITLE
Does ETL box handle large datasets & how DbMerge works

### ISSUE THREAD

**Issue by Thwaitesy** [roadrunnerlenny/etlbox#51]
**Is your feature request related to a problem? Please describe.**
It's more a question, but can be translated into some documentation for others. 

I have a sql table with 20 million records - can ETLBox handle datasets this big? 

If so - are there any considerations that we should take into account to make things more efficient?

**Comment by roadrunnerlenny** [comment]
Hi,

Thank you for your feedback and question about ETLbox.

The short answer is: ETLBox as an ETL tool was designed to handle big amounts of data and therefore you shouldn’t encounter any issues working with 20 millions of rows. 

Here some details how ETLbox works under the hood:

A basic dataflow consists of a source, a transformation and a destination.

The source will start reading and post its data into the components connected to its output. As soon as a connected component retrieves any data in its input, the component will start with processing the data and then send it further down the line to its connected components. The dataflow will finish when all data from the source(s) are read and received from the destinations.

Of course, all data is processed asynchronously by the components. Each compoment has its own set of buffers, so while the source is still reading data, the transformations can already process it and the destinations can start writing the processed information into their target.

Therefore (depending on your transformation) your 20 mio rows don’t even have to fit into the memory - only as much data is stored in memory as you need to keep the buffers filled so that there is always enought data available to write into the destination. 

If you are writing into a DBDestination, data is inserted using bulk inserts. Depending on your database, there are some tweaks on your database that you can enable to improve bulk insert performance. Which database are you using?

There are some transformations (e.g.) which are blocking transformations. These transformations will need more memory because they can only work properly when all data was loaded into the memory. E.g. a sort can only sort your data when it knows all records.
You can read more about it here: https://etlbox.net/articles/dataflow/dataflow_transformations.html

Does that answer your question?

**Comment by Thwaitesy** [comment]
Ok - it's good to know it will handle that volume.

A couple more questions sorry :-)

Our scenario is that we have a transactions table that is growing rapidly that we need to sync and transform it every night to another database for various reasons (Both db's are using managed Azure SQL by the way). Our transactions table has a "LastUpdated" field which we are using for comparing and it also has an "Exclude" field to indicate that we shouldn't sync the transaction ever.

Our current setup is to run a DBMerge with a predicate inside the LinkTo that excludes rows or not based on the "Exclude" field.

```
var sourceDatabase = new DBSource<TSourceEntity>(new SqlConnectionManager(this._sourceConnection), strategy.SourceTableName);

var destinationDatabase = new DBMerge<TDestinationEntity>(new SqlConnectionManager(this._destinationConnection), strategy.DestinationTableName);

//The TransformRow here is just a mapping layer from source to dest.. 
var transform = new RowTransformation<TSourceEntity, TDestinationEntity>(source => strategy.TransformRow(source));

//If we exclude a row, it needs to go into the void... 
sourceDatabase.LinkTo(transform, (rowToSync) => !rowToSync.Exclude, (rowToVoid) => rowToVoid.Exclude);

transform.LinkTo(destinationDatabase);

sourceDatabase.Execute();
destinationDatabase.Wait();
```

```
//Note: we also use this DestinationEntity as the entity in EFCore to build the destination db.... 
public class Transaction : MergeableRow
{
        [Key]
        [IdColumn]
        public Guid TransactionId { get; set; }

        [CompareColumn]
        public DateTimeOffset LastUpdated { get; set; }

        + a few more in here...
}
```

My two questions are -

1) With a DBMerge it uses a lookup table - in the docs you point out that if there were 20 million rows, it would need to store the lookup table in memory for those 20 million rows. This obviously would put a bit of strain on things.. so how would you go about making this more efficient? Do we utilise the DBSource SQL thing where we only select transactions that have last changed since we last run a successful sync? This is ok for gradually growing tables.. but if we started with 0 in the destination the sync would still be in a world of pain. 

- Could we use block transform and process 50k rows at a time?
- Would it be easier to not use DBMerge and just truncate and sync the table each time?
- Is 20 million items with 3 values in an array like 100mb and we shouldn't worry about the memory costs here?

2) We are struggling with merging 20k rows... which I don't think is a performance thing, I think it's an implementation issue of how we are doing it.. (I hope)

At 1,000 transactions everything syncs perfectly, at 20,000 transactions it seems to delete all transactions then only sync some back over or could be determining that some should be deleted and some should be added... the number of rows synced is different each time we run the process and we have never got the same count between the tables.. we are always about half or a third of the table size out...

- Could this have something to do with using the predicates in the LinkTo? Should we use the DBSource SQL thing and exclude stuff from the sync that way? or is it fine to use the predicate the way we have?

- The fact it's deleting the txns to me indicates that the [IdColumn] or the [CompareColumn] isnt working as expected and its deciding to delete all rows for some reason

- Could the Guid or the DateTimeOffset be having any effect here? Im thinking maybe there is a precision on the DateTimeOffset that may not be the same between the two so that might cause us to always sync.. but it shouldnt cause us to delete some and sync some...

- Reading your docs I think my next step here is to create a DeltaTable and push the information there to see why its deciding to do crazy things :-)

Anyway - Ill keep fluffing around as im keen to get this going because I can see some awesome benefits in using an in process ETL so we can transform things easily as opposed to the other options.

Would love your insights :-)

Cheers,
Sam

**Comment by anlennar** [comment]
Hi Sam,

thanks for the insights how you want to use ETLBox. I think the scenario you describe sounds very feasible and I think it should work with ETLBox. 

Let me try to answer your questions: 
First of all, the use of the predicate for the exclude looks good and should work as expected. I don't think that you will run in problems here. The issues you are facing is probably related to the Merging-approach.

So basically you have a transaction table, where you can identify the  latest transactions with an update timestamp.First questions here: Is there also some way to identify deletions, or do you want to take deletions into consideration at all?

 I think you now have 2 options here.

Option 1: You always load the full data from the source (actually then the update timestamp is of no use), and then you copy it into the destination. If you are not interesting in the changes between source and destination, the easiest way would be to just truncate your target table and then reinsert everything again. That way just copy data from A to B, without any necessarily to check the data in the target table before inserted. I would call this a "Full" copy of your data. At first glance this does sound like the slower approach, but actually it can depend on the amount of updates that happen on your source transaction table. E.g., if every transaction gets updated every day, this could become actually the fastest approach.  Other than that, you could improve the speed of this process by using some kind of partitioning. E.g. if you know that all transactions from the previous year(s) are never get updated again, you could create a table that is partitioned in your destination by year. This way you would always leave the data from the previous years untouched and just copy the current year always in full. To implement this scenarion, you just would use:
DbSource --> RowTransformation --> DbDestination
and a SqlTask that would remove the data in the destination before the copying. 

Option 2: You use the update timestamp to identify the changes. First of all, you need to remember the last-import timestamp somewhere in your code or database. Then, you could create a delta load using the DbSource combined with a sql statement like this:
```
DateTime lastUpdateTime = GetLastUpdateDate();
DBSource source = new DbSource() {
    ConnectionManager = new SqlConnectionManager("..."),
    Sql = $"SELECT * FROM transactions WHERE updatetimestamp > {lastUpdateTime.ToString("yyy-MM-dd"}"
}
```
So you would only receive the changes here. 
Next step would be using a RowTransformation (to transform the objects and filter out the records with the exclude-flag). 
Now your data will arrive at the DbMerge. Actually, a DbMerge is not the only option you have here. As you are using Azure Sql and you can use T-SQL, you just could copy your data in a temp table, e.g. "deltaTransactions". Always truncate this table before you add new data. Now you could use the MERGE-statement from T-SQL to merge the data from the "deltaTransactions" table into your destination table. The Merge operation would then be done on your target server.

But this won't work on other databases as they don't offer a MERGE or UPSERT statement. Also, you probably won't to avoid having T-Sql code in your application. 
 That's why I create the DbMerge as destination. And it should actually work as example in the example code you posted.
 Let me give you a quick explanation how the DbMerge works under the hood: First of all, it will *always* load all data from the target table into the memory. So in your case it will always load all transaction from the destinations into memory before starting the merge. This could be a downside, but I would say that 20 mio transactions doesn't sound too much. After having read all data from the source, it will then delete the missing records and *the records that needs to be updated* using a "DELETE" statement. This will work find if there are only a few deletes to be made. But it could be that the amounts of deletions are very high, and the delete statement can become slow. DBMerge offers the option to enforce a TRUNCATE instead. Just set the property `UseTruncateMethod` to true. This will work faster. (You could loose some information about the updates, but in your case you are not interested in these) 
After having the records deleted that are missing or needed to be updated, the new and updated records are inserted into the target using a bulk insert. 

So far, so good. 
In your case, it could be that there are issues with the use of DateTimeOffset. I haven't used that before in my tests, but I can write a test to check if this does behave differently. Also, it could be that the DELETE-statement inside the DbMerge does cause some issues. it could be that it has problems with a large amounts of Deletions. I would also need to write a test to check if this could the root cause. In the meanwhile, you could try to use the DBMerge with setting the property `UseTruncateMethod` to true.
```var merge = new DBMerge(connMan, tableName) {
  UseTruncateMethod = true
};
```
If this is solving the issue, it would be an indicator that there is something wrong with the delete database.

Also one thing which is interesting is how you store the last update - date in your database. Is it of type DateTime? 
Based on your example code, I should be able to create a test the reproduces your scenario. Perhaps I can reproduce the bug in my test environment. If I understand you right, things only go wrong when you start working with a higher amount of data (namely > 20.000 rows)?

**Comment by Thwaitesy** [comment]
Ok thanks for this - I have investigated a couple of points you mentioned and I think we are a bit closer to figuring out what is going on.

I created a test that does a merge of 900 rows, and one with 1100 rows and the UseTruncateMethod does work as you suggested 👍  

```
UseTruncateMethod = false
900 rows = PASS
1100 rows = FAIL (It results in 1039 rows...)

UseTruncateMethod = true
900 rows = PASS
1100 rows = PASS
```

Here is a log of the process - At the end there I write out the count of how many "I", "E", "D" values are in the delta table.. which it shows Existing = 1119.. which is odd as there is only 1100 rows in the table.. 🤷‍♂️

![image](https://user-images.githubusercontent.com/1551204/74582407-fc3b6c00-5006-11ea-9691-43906227c2ce.png)

Also - I have migrated the LastUpdated column to be DateTime instead of DateTimeOffset and it doesn't have an impact on the outcome of the tests.

Unfortunately due to SQL permission restrictions we can't give Truncate permissions to the user running this process, so I'm keen to dig into the code at some point over the weekend and try see if I can find where things are going south.

For now, we will run the SQLTask you mentioned to delete all rows in the table and do a straight sync instead of a DBMerge 👍 

Thanks for your help so far on this - appreciate it

**Comment by Thwaitesy** [comment]
Here is the passing test logs for comparison...

![image](https://user-images.githubusercontent.com/1551204/74582603-eed3b100-5009-11ea-97bd-b9d48f853a76.png)

**Comment by roadrunnerlenny** [comment]
Hi @Thwaitesy ,
thanks again for your feedback and the logs! That is actually a quite good description of the bug, I will try to reproduce this in my test environment. I'll keep you updated if I was able to make some progress here!

**Comment by roadrunnerlenny** [comment]
Hi again,

a quick update on this. I was able to reproduce this bug within my tests. There was a flaw with the sequence of inserting incoming data and when to delete missing records. In the current implementation, the DBMerge tried to delete missing records before every batch write. This worked fine with small amounts of data (< 1000 rows, as this is the default batch size), but could break if the data exceeded the batch size. (But not only by 1, it must have been actually a little bit more - it happened when data was prepared for bulk insert at the same time while the missing records where identified).
The solution was to move the deletion of the missing records at the end of the complete DbMerge operation. I should have fix edthis now. The fix will be part of the next release. 

The next version is not fully ready yet, but I already created a pre-release version which contains the bugfix:
https://www.nuget.org/packages/ETLBox/1.8.0-rc2

Please let me know if this pre-release version works for you. 

*Note: This release does contain some minor breaking changes: I adjusted the names for CSVSource/CSVDestination and DBSource/DBDestination/DBMerge to real CamelCase - they are now CsvSource/CsvDestination and DbSource/DbDestination/DBMerge. ConnectionString was renamed to SqlConnectionString.*

**Comment by roadrunnerlenny** [comment]
Hi @Thwaitesy ,

did the pre-release version work for you?

**Comment by roadrunnerlenny** [comment]
Hi @Thwaitesy , 

I just released version 1.8.0 (https://www.nuget.org/packages/ETLBox/1.8.0) that contains the fix for your issue.

**Comment by Thwaitesy** [comment]
Hi @roadrunnerlenny,

We havent been able to update yet - We will do so over the next couple of weeks and will let you know how we go.

Thanks for the fix :-)
